File: 3.0/python/12.py
stack = []
ln = 0

for i in input():
    if i in "([{":
        stack.append(i)
        ln += 1
    else:
        if ln == 0:
            ln = -1
            break
        else:
            if i == ")" and stack[-1] == "(":
                stack.pop()
                ln -= 1 
            elif i == "]" and stack[-1] == "[":
                stack.pop()
                ln -= 1 
            elif i == "}" and stack[-1] == "{":
                stack.pop()
                ln -= 1 
            else:
                ln = -1
                break

if ln == 0:
    print("yes")
else:
    print("no")

File: 3.0/golang/12.go
package main

import (
	"fmt"
	"strings"
)

type stack struct {
	arr []rune
	ln  int
}

func (stack *stack) append(x ...rune) {
	for _, v := range x {
		stack.arr = append(stack.arr, v)
		stack.ln += 1
	}
}

func (stack *stack) pop() rune {
	c := stack.arr[stack.ln-1]
	stack.arr = stack.arr[:stack.ln-1]
	stack.ln -= 1

	return c
}

func (stack *stack) last() rune {
	return stack.arr[stack.ln-1]
}

func main() {
	var slice []rune
	stack := stack{arr: slice, ln: 0}

	var input string
	fmt.Scanln(&input)

Loop:
	for _, v := range input {
		if strings.ContainsRune("([{", v) {
			stack.append(v)
		} else {
			if stack.ln == 0 {
				stack.ln = -1
				break
			} else {
				switch {
				case v == ')' && stack.last() == '(':
					stack.pop()
				case v == ']' && stack.last() == '[':
					stack.pop()
				case v == '}' && stack.last() == '{':
					stack.pop()
				default:
					stack.ln = -1
					break Loop
				}
			}
		}
	}

	if stack.ln == 0 {
		fmt.Println("yes")
	} else {
		fmt.Println("no")
	}
}
